galaxy_ng/app/api/ui/views/ai_index.py

```diff
@@ -0,0 +1,180 @@
+    extend_schema,
+    inline_serializer
+)
+from rest_framework import status
+from rest_framework.response import Response
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError as DRFValidationError
+
+
+from galaxy_ng.app.access_control import access_policy
+from galaxy_ng.app.api import base as api_base
+from galaxy_ng.app.api.v1.models import LegacyNamespace
+from galaxy_ng.app.models.aiindex import AIIndexDenyList
+from galaxy_ng.app.models.namespace import Namespace
+
+NamespaceOrLegacyNamespace = (
+    access_policy.NamespaceAccessPolicy | access_policy.LegacyAccessPolicy
+)  # DRF 3+ allows composed access policies using & or |
+
+
+class AIIndexBaseView(api_base.APIView):
+    permission_classes = [NamespaceOrLegacyNamespace]
+
+    def _verify_permission_for_namespace_object(self, request, scope, reference):
+        """Checks if the user has permission to modify the namespace object."""
+
+        model = {"namespace": Namespace, "legacy_namespace": LegacyNamespace}
+        try:
+            namespace = model[scope].objects.get(name=reference)
+        except ObjectDoesNotExist:
+            raise DRFValidationError(f"Referenced {scope} {reference!r} does not exist")
+        except KeyError:
+            raise DRFValidationError(f"Invalid scope {scope!r}")
+
+        # This is a method on DRF that will use permission_classes to check
+        # for object permissions, because this is not a ModelViewSet, this method
+        # must be called manually.
+        self.check_object_permissions(request, namespace)  # will raise permission error
+
+
+class AIIndexAddView(AIIndexBaseView):
+    """Allows to add a namespace to AIIndexDenyList."""
+    action = "ai-index-add"
+
+    @extend_schema(
+        request=inline_serializer(
+            name="AddToAIIndexDenyList",
+            fields={"reference": serializers.CharField()}
+        ),
+        parameters=[
+            OpenApiParameter(
+                "scope", OpenApiTypes.STR, "path", enum=["namespace", "legacy_namespace"]
+            )
+        ],
+        responses={
+            201: inline_serializer(
+                name="AddedToAIIndexDenyList",
+                fields={"scope": serializers.CharField, "reference": serializers.CharField},
+            ),
+            403: "Forbidden",
+            400: "Bad Request",
+            409: "Conflict"
+        },
+    )
+    @transaction.atomic
+    def post(self, request, scope, *args, **kwargs):
+        """Adds a collection to the AI Index deny list.
+
+        http::
+            POST _ui/v1/ai_index/{namespace|legacy_namespace}/
+            {
+                "reference": "some_name"
+            }
+
+        responses:
+            201: Created
+            400: Bad Request (invalid payload data)
+            409: Conflict (data already in the model)
+        """
+        data = {"scope": scope, "reference": request.data.get("reference")}
+        obj = AIIndexDenyList(**data)
```

Comment:
Should we attempt to cast this across both namespace and legacynamespace if they have matching names or if legacynamespace has a foreignkey relationship with the namespace?

Reply:
I am not sure, if namespace and legacy_namespace have matching names, is it owned by the same owners?
How is the relationship between legacy and namespace?

Should I also add this here? I can duplicate the entry